my/core/query.py

```diff
@@ -0,0 +1,386 @@
+                    tpically passed directly to 'sorted'
+
+    order_key:      a string which represents a dict key or attribute name
+                    to use as they key to sort by
+
+    order_value:    predicate which determines which attribute on an ADT-like item to sort by,
+                    when given its value. lambda o: isinstance(o, datetime) is commonly passed to sort
+                    by datetime, without knowing the attributes or interface for the items in the src
+
+    default:        while ordering, if the order for an object cannot be determined,
+                    use this as the default value
+
+    reverse:        reverse the order of the resulting iterable
+
+    limit:          limit the results to this many items
+
+    drop_errors:    ignore any errors from the src
+
+    raise_errors:   raise errors when recieved from the input src
+    """
+
+    it: Iterable[ET] = []  # default
+    # check if this is a locator
+    if type(src) == tuple and len(src) == 2:  # type: ignore[arg-type]
+        it = locate_function(src[0], src[1])()  # type: ignore[index]
+    elif callable(src):
+        # hopefully this returns an iterable and not something that causes a bunch of lag when its called?
+        # should typically not be the common case, but giving the option to
+        # provide a function as input anyways
+        it = src()
+    else:
+        # assume it is already an iterable
+        if not isinstance(src, Iterable):
+            low(f"""Input was neither a locator for a function, or a function itself.
+Expected 'src' to be an Iterable, but found {type(src).__name__}...
+Will attempt to call iter() on the value""")
+        it = src
+
+    # try/catch an explicit iter() call to making this an Iterator,
+    # to validate the input as something other helpers here can work with,
+    # else raise a QueryException
+    try:
+        itr: Iterator[ET] = iter(it)
+    except TypeError as t:
+        raise QueryException("Could not convert input src to an Iterator: " + str(t))
+
+    # if both drop_errors and raise_errors are provided for some reason,
+    # should raise errors before dropping them
+    if raise_errors:
+        itr = _raise_errors(itr)
+
+    if drop_errors:
+        itr = _drop_errors(itr)
+
+    if where is not None:
+        itr = filter(where, itr)
+
+    if order_by is not None or order_key is not None or order_value is not None:
+        # we have some sort of input that specifies we should reorder the iterator
+
+        order_by_chosen: Optional[OrderFunc] = order_by  # if the user just supplied a function themselves
+        if order_by is None:
+            # https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.spy
+            [first_item], itrc = more_itertools.spy(itr)
+            # replace the 'itr' in the higher scope with itrc -- itr is consumed by more_itertools.spy
+            itr = itrc
```

Comment:
I usually just write `[first_item], itr = spy(itr)`. But guess it might be cryptic to someone who sees spy for the first time :)